R/summarize.R

```diff
@@ -1,29 +1,85 @@
 #' @export
-#' @importFrom dplyr summarise_
+#' @importFrom dplyr bind_cols summarise_ select
+#' @importFrom purrr map_chr map_df
 #' @importFrom crunch crtabs
 summarise_.CrunchDataset <- function (.data, ..., .dots) {
     dots <- all_dots(.dots, ..., all_named = TRUE)
-    fmla <- dots_to_formula(dots, groups(.data))
-    out <- as_tibble(crtabs(fmla, data=.data))
-    ## Sort the return columns based on the request order
-    out <- out[, c(setdiff(names(out), names(dots)), names(dots))]
+    unweighted <- dots %>% map_chr(~as.character(.$expr)[[1]]) == "unweighted_n"
+    counts <- dots[unweighted]
+    non_counts <- dots[!unweighted]
+    fmla <- dots_to_formula(non_counts, groups(.data))
+    
+    # When there are no groups or summary functions, we can't 
+    # use crtabs to get a cube, but unweighted_n() is equivalent to nrow(ds)
+    # Otherwise we can use crtabs to get the cube and extract the unweighted counts
+    # from the cube. This doesn't work for weighted counts because they 
+    # are not included in some cubes. So in that case we rely on the the server
+    # `n` function. 
+    if (length(non_counts) == 0 && length(groups(.data)) == 0) {
+        out <- map_df(counts, ~nrow(.data))
+    } else {
+        cube <- crtabs(fmla, data=.data)
```

Comment:
IMO this section would read more cleanly as:

```
out <- as_tibble(crtabs(fmla, data=.data))
# Comment about map_df
unweighted_n <-  map_df(counts, ~ out$row_count)
out$row_count <- NULL
out <- bind_cols(out, unweighted_n)
```

Avoids the `@` into the cube and looks more explicit to me.